### PR TITLE
ARROW-7631: [C++][Gandiva] return zero if there is an overflow while downscaling a decimal

### PR DESCRIPTION
**Pull request by pprudhvi**
- also add castDecimalNullOnOverflow function

### CODE REVIEW

File: cpp/src/gandiva/precompiled/decimal_wrapper.cc [hunk 1]
```diff
@@ -229,14 +229,43 @@ void castDECIMAL_float32(float in, int32_t x_precision, int32_t x_scale,
 }
 
 FORCE_INLINE
-void castDECIMAL_decimal128(int64_t x_high, uint64_t x_low, int32_t x_precision,
-                            int32_t x_scale, int32_t out_precision, int32_t out_scale,
-                            int64_t* out_high, int64_t* out_low) {
+boolean castDecimal_return_overflow(int64_t x_high, uint64_t x_low, int32_t x_precision,
```
Comment (pravindra) [inline cpp/src/gandiva/precompiled/decimal_wrapper.cc]: can you please name it as castDecimal_internal ? having the return value in name looks odd.

File: cpp/src/gandiva/precompiled/decimal_ops.cc [hunk 2]
```diff
@@ -544,7 +544,10 @@ static BasicDecimal128 ModifyScaleAndPrecision(const BasicDecimalScalar128& x,
     return x.value().IncreaseScaleBy(delta_scale);
   } else {
     // Do not do any rounding, that is handled by the caller.
-    return x.value().ReduceScaleBy(-delta_scale, false);
+    auto result = x.value().ReduceScaleBy(-delta_scale, false);
+    DECIMAL_OVERFLOW_IF(BasicDecimal128::Abs(result) > GetMaxValue(out_precision),
```
Comment (pravindra) [inline cpp/src/gandiva/precompiled/decimal_ops.cc]: please modify decimalops::Convert to not add delta if there is an overflow.

File: cpp/src/gandiva/precompiled/decimal_wrapper.cc (hunk 1 above)
Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/decimal_wrapper.cc]: done

File: cpp/src/gandiva/precompiled/decimal_ops.cc (hunk 2 above)
Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/decimal_ops.cc]: done

**Review by pravindra — APPROVED** [review]

**Pull request closed without merging**